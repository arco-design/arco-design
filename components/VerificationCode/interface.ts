import { CSSProperties, ReactNode } from 'react';
import { InputProps } from '../Input';

/**
 * @title VerificationCode
 * @zh `2.55.0` 支持
 * @en 2.55.0
 */
export interface VerificationCodeProps {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 默认值
   * @en The initial input content
   */
  defaultValue?: string;
  /**
   * @zh 验证码输入框的值，受控模式
   * @en The input content value
   */
  value?: string;
  /**
   * @zh 验证码的长度，根据长度渲染对应个数的输入框
   * @en The length of the verification code, rendering the corresponding number of input boxes according to the length
   * @defaultValue 6
   */
  length?: number;
  /**
   * @zh 尺寸
   * @en Size
   */
  size?: InputProps['size'];
  /**
   * @zh 是否是密码模式
   * @en Password mode
   */
  masked?: boolean;
  /**
   * @zh 禁用
   * @en disabled
   */
  disabled?: boolean;
  /**
   * @zh 只读
   * @en readOnly
   */
  readOnly?: boolean;
  /**
   * @zh 状态
   * @en Status
   */
  status?: 'error';
  /**
   * @zh 校验函数，用户输入值改变时触发
   * @en Verification function, triggered when the user input value changes
   *
   */
  validate?: (data: { inputValue: string; value: string; index: number }) => boolean | string;
  /**
   * @zh 分隔符。可在不同索引的输入框后自定义渲染分隔符
   * @en Separator. Customizable rendering separators after input boxes with different indexes
   */
  separator?: (data: { index: number; character: string }) => ReactNode;
  /**
   * @zh 输入值改变时触发的回调
   * @en Callback triggered when input value changes
   */
  onChange?: (value: string) => void;
  /**
   * @zh 输入框都被填充后触发的回调
   * @en A callback triggered after the input boxes are filled in
   */
  onFinish?: (value: string) => void;
}
