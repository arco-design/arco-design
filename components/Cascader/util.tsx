import { isArray } from '../_util/is';
import Store from './base/store';
import { CascaderProps } from './interface';

export const ValueSeparator = '__arco_cascader__';

export const SHOW_PARENT = 'parent';
export const SHOW_CHILD = 'child';

export const PANEL_MODE = {
  cascader: 'cascader',
  select: 'select',
};

export function isEmptyValue(value) {
  return !value || (isArray(value) && value.length === 0);
}

export function getConfig(props: CascaderProps) {
  return {
    showEmptyChildren: props.showEmptyChildren,
    changeOnSelect: props.changeOnSelect,
    lazyload: !!props.loadMore,
    fieldNames: props.fieldNames,
    filterOption: props.filterOption,
    showParent:
      props.mode === 'multiple' && !props.changeOnSelect && props.checkedStrategy === SHOW_PARENT,
  };
}

export function getStore(props, value) {
  const tmp = value ? (Array.isArray(value[0]) ? value : [value]) : [];
  return new Store(props.options || [], tmp, getConfig(props));
}

export const transformValuesToSet = (values: string[][]) => {
  const _values = values || [];
  const valuesSet = _values.reduce((set, next) => {
    // 'next' could be a string.
    set.add([].concat(next).join(ValueSeparator));
    return set;
  }, new Set());

  return valuesSet;
};

export const valueInSet = (set, value: string[]) => {
  const _value = value || [];
  return set.has(_value.join(ValueSeparator));
};

export const removeValueFromSet = (set, value: string[]) => {
  const _value = value || [];
  return set.delete(_value.join(ValueSeparator));
};

export const formatValue = (value, isMultiple, store?): string[][] | undefined => {
  let _value = [];
  if (value === undefined) {
    _value = [];
  } else if (isMultiple) {
    _value = value;
  } else {
    _value = [value];
  }

  if (store && store.config.showParent) {
    const checkedNodes = store.getCheckedNodes();
    const valuesSet = transformValuesToSet(checkedNodes.map((node) => node.pathValue));
    const result = [];
    const temp = {};
    _value.map((v) => {
      v.some((_, index, arr) => {
        const curVal = arr.slice(0, index + 1);
        const pass = valueInSet(valuesSet, curVal);
        if (pass && !temp[curVal.join(ValueSeparator)]) {
          result.push(curVal);
          temp[curVal.join(ValueSeparator)] = 1;
        }
        return pass;
      });
    });

    return result;
  }
  return _value;
};

// change check status to false
const deny2Checked = (option) => {
  const deny = (options) => {
    return !Array.isArray(options)
      ? false
      : options.every((item) => {
          if (item._checked || item.disabled) {
            return true;
          }
          return deny(item.children);
        });
  };
  return option._halfChecked && deny(option?.children);
};

export const getMultipleCheckValue = (propsValue, store: Store<any>, option, _checked) => {
  const checked = _checked && deny2Checked(option) ? false : _checked;

  const beforeValueSet = store.getCheckedNodes().reduce((set, node) => {
    set.add(node.pathValue.join(ValueSeparator));
    return set;
  }, new Set());

  option.setCheckedState(checked);
  const checkedNodes = store.getCheckedNodes();
  const currentValue = checkedNodes.map((node) => node.pathValue);
  const currentValueSet = transformValuesToSet(currentValue);

  const newValueSet = new Set();
  return propsValue
    .filter((v) => {
      // v 不在 beforeValueSet 中，说明 v 不包含对应的option。直接返回true，不应该清除掉。
      if (!valueInSet(beforeValueSet, v) || valueInSet(currentValueSet, v)) {
        newValueSet.add(v.join(ValueSeparator));
        return true;
      }
    })
    .concat(
      currentValue.filter((v) => {
        return !valueInSet(newValueSet, v);
      })
    );
};
