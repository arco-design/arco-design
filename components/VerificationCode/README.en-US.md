`````
Component / Data Entry

# VerificationCode

VerificationCode
`````

%%Content%%

## API

### VerificationCode

2.55.0

|Property|Description|Type|DefaultValue|
|---|---|---|---|
|disabled|disabled|boolean |`-`|
|masked|Password mode|boolean |`-`|
|readOnly|readOnly|boolean |`-`|
|length|The length of the verification code, rendering the corresponding number of input boxes according to the length|number |`6`|
|defaultValue|The initial input content|string |`-`|
|value|The input content value|string |`-`|
|size|Size|[InputProps](input#input)['size'] |`-`|
|status|Status|'error' |`-`|
|className|Additional css class|string \| string[] |`-`|
|style|Additional style|CSSProperties |`-`|
|onChange|Callback triggered when input value changes|(value: string) => void |`-`|
|onFinish|A callback triggered after the input boxes are filled in|(value: string) => void |`-`|
|separator|Separator. Customizable rendering separators after input boxes with different indexes|(data: { index: number; character: string }) => ReactNode |`-`|
