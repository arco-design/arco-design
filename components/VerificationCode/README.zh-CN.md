`````
组件 / 输入

# 验证码输入 VerificationCode


验证码输入组件，`2.55.0` 支持
`````


%%Content%%

## API

### VerificationCode

`2.55.0` 支持

|参数名|描述|类型|默认值|
|---|---|---|---|
|disabled|禁用|boolean |`-`|
|masked|是否是密码模式|boolean |`-`|
|readOnly|只读|boolean |`-`|
|length|验证码的长度，根据长度渲染对应个数的输入框|number |`6`|
|defaultValue|默认值|string |`-`|
|value|验证码输入框的值，受控模式|string |`-`|
|size|尺寸|[InputProps](input#input)['size'] |`-`|
|status|状态|'error' |`-`|
|className|节点类名|string \| string[] |`-`|
|style|节点样式|CSSProperties |`-`|
|onChange|输入值改变时触发的回调|(value: string) => void |`-`|
|onFinish|输入框都被填充后触发的回调|(value: string) => void |`-`|
|separator|分隔符。可在不同索引的输入框后自定义渲染分隔符|(data: { index: number; character: string }) => ReactNode |`-`|
|validate|校验函数，用户输入值改变时触发|(data: { inputValue: string; value: string; index: number }) => boolean \| string |`-`|
