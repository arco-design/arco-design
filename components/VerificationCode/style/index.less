@import '../../Input/style/token.less';
@import './token.less';

@verification-code-prefix-cls: ~'@{prefix}-verification-code';
@input-prefix-cls: ~'@{prefix}-input';

.@{verification-code-prefix-cls} {
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  column-gap: 4px;

  .@{input-prefix-cls} {
    width: @input-size-default-height;
    padding-left: 0;
    padding-right: 0;
    text-align: center;
  }

  .@{input-prefix-cls}-size-small {
    width: @input-size-small-height;
  }

  .@{input-prefix-cls}-size-mini {
    width: @input-size-mini-height;
  }

  .@{input-prefix-cls}-size-large {
    width: @input-size-large-height;
  }
}

.@{verification-code-prefix-cls}-rtl {
  direction: rtl;
}
