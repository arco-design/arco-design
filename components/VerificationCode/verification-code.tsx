import React, { forwardRef, useRef, useContext, Fragment } from 'react';
import cs from '../_util/classNames';
import InputComponent from '../Input/input-element';
import { ConfigContext } from '../ConfigProvider';
import useMergeProps from '../_util/hooks/useMergeProps';

import { VerificationCodeProps } from './interface';
import useVerificationCode from './hooks/useVerificationCode';

const defaultProps = {
  length: 6,
};

export function VerificationCodeComponent(baseProps: VerificationCodeProps, _) {
  const ctx = useContext(ConfigContext);
  const props = useMergeProps(baseProps, defaultProps, ctx.componentConfig?.VerificationCode);

  const { size, separator, status, length, masked, disabled } = props;
  const focusEleRefList = useRef([]);
  const { value, filledValue, getInputProps } = useVerificationCode({
    defaultValue: props.defaultValue,
    value: props.value,
    length,
    getInputRefList: () => focusEleRefList.current,
    onChange: props.onChange,
    onFinish: props.onFinish,
  });

  const prefix = ctx.getPrefixCls('verification-code');
  const prefixInput = ctx.getPrefixCls('input');

  return (
    <div
      className={cs(`${prefix}`, props.className, {
        [`${prefix}-rtl`]: ctx.rtl,
      })}
      style={props.style}
    >
      {filledValue.map((v, index) => {
        const { onChange, onClick, onPaste, onKeyDown, ...restInputProps } = getInputProps(index);
        return (
          <Fragment key={index}>
            <InputComponent
              disabled={props.disabled}
              readOnly={props.readOnly}
              className={cs(prefixInput, `${prefix}-input`, {
                [`${prefixInput}-size-${size}`]: size,
                [`${prefixInput}-${status}`]: status,
                [`${prefixInput}-disabled`]: disabled,
                [`${prefixInput}-rtl`]: ctx.rtl,
              })}
              ref={(node) => {
                focusEleRefList.current[index] = node?.dom;
              }}
              {...restInputProps}
              onClick={!props.readOnly ? onClick : undefined}
              onPaste={!props.readOnly ? onPaste : undefined}
              onKeyDown={!props.readOnly ? onKeyDown : undefined}
              onChange={
                !props.readOnly
                  ? (inputValue) => {
                      if (props.validate) {
                        const result = props?.validate({ inputValue, index, value });
                        if (result !== false) {
                          onChange(typeof result === 'string' ? result : inputValue);
                        }
                      } else {
                        onChange(inputValue);
                      }
                    }
                  : undefined
              }
              type={masked ? 'password' : 'text'}
            />
            {separator?.({ index, character: v })}
          </Fragment>
        );
      })}
    </div>
  );
}

const VerificationCode = forwardRef(VerificationCodeComponent);

export default VerificationCode;
