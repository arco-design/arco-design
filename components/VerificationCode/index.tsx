import VerificationCode from './verification-code';
import { VerificationCodeProps } from './interface';

export { VerificationCodeProps };

export default VerificationCode;
