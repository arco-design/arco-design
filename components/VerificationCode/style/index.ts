import '../../style/index.less';
import '../../Input/style';
import './index.less';
