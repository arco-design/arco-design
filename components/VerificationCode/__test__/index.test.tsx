import React from 'react';
import componentConfigTest from '../../../tests/componentConfigTest';
import VerificationCode from '..';
import { fireEvent, render } from '../../../tests/util';

componentConfigTest(VerificationCode, 'VerificationCode');

describe('Alert', () => {
  it('control correctly', () => {
    let value = '123';

    const Demo = () => (
      <VerificationCode
        value={value}
        onChange={(v) => {
          value = v;
        }}
      />
    );

    const component = render(Demo());

    const inputs = component.querySelectorAll('input');
    inputs.forEach((input, index) => {
      expect(input.getAttribute('value')).toBe(value.split('')[index] || '');
    });

    fireEvent.change(inputs.item(3), { target: { value: 4 } });
    expect(value).toBe('1234');
    component.rerender(Demo());

    fireEvent.change(component.querySelectorAll('input').item(4), { target: { value: 5 } });
    expect(value).toBe('12345');

    component.rerender(Demo());

    fireEvent.change(component.querySelectorAll('input').item(5), { target: { value: 6 } });

    expect(value).toBe('123456');
  });
});
