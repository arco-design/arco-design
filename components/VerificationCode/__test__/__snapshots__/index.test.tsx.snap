// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.VerificationCode default 1`] = `
<div
  class="arco-verification-code"
>
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
</div>
`;

exports[`ConfigProvider componentConfig.VerificationCode set className globally 1`] = `
<div
  class="arco-verification-code global-default-class"
>
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
</div>
`;

exports[`ConfigProvider componentConfig.VerificationCode set className globally and component set className itself. 1`] = `
<div
  class="arco-verification-code component-class"
>
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
</div>
`;

exports[`ConfigProvider componentConfig.VerificationCode set data-* property 1`] = `
<div
  class="arco-verification-code"
>
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
  <input
    class="arco-input arco-verification-code-input"
    type="text"
    value=""
  />
</div>
`;
